### PR TITLE
Serve DestinationRule file mounts over SDS

### PR DESCRIPTION
**Pull request by howardjohn**
For https://github.com/istio/istio/issues/21833

This covers DR, Gateway will probably be added to this PR later

### CODE REVIEW

**Comment by howardjohn** [comment]
I don't think we have any tests for this either

**Comment by howardjohn** [comment]
#22052 adds tests here

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -963,6 +963,19 @@ func applyUpstreamTLSSettings(opts *buildClusterOpts, tls *networking.TLSSetting
 					},
 				},
 			}
+		} else if tls.Mode == networking.TLSSettings_MUTUAL {
+			// These are certs being mounted from within the pod. Rather than reading directly in Envoy,
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: what about the ca certs?

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: I need to add this, good catch.

/hold

should probably also decide if we can do https://github.com/envoyproxy/envoy/issues/10387 before investing time in this

**Comment by howardjohn** [comment]
this is ready for review

File: pilot/pkg/model/authentication.go [hunk 2]
```diff
@@ -252,3 +254,41 @@ func getConfigsForWorkload(configsByNamespace map[string][]Config,
...
+	PrivateKeyPath    string
+	CaCertificatePath string
+}
+
+// GetResourceName converts a SdsCertificateConfig to a string to be used as an SDS resource name
+func (s SdsCertificateConfig) GetResourceName() string {
+	return fmt.Sprintf("file-cert:%s~%s", s.CertificatePath, s.PrivateKeyPath)
+}
+
+// GetResourceName converts a SdsCertificateConfig to a string to be used as an SDS resource name for the root
```
Comment (ramaraochavali) [inline pilot/pkg/model/authentication.go:269]: nit: /s/GetResourceName/GetRootResourceName

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 3]
```diff
@@ -1026,6 +1026,23 @@ func applyUpstreamTLSSettings(opts *buildClusterOpts, tls *networking.ClientTLSS
...
+				PrivateKeyPath:    model.GetOrDefault(proxy.Metadata.TLSClientKey, tls.PrivateKey),
+				CaCertificatePath: model.GetOrDefault(proxy.Metadata.TLSClientRootCert, tls.CaCertificates),
+			}
+			tlsContext.CommonTlsContext.TlsCertificateSdsSecretConfigs = append(tlsContext.CommonTlsContext.TlsCertificateSdsSecretConfigs,
+				authn_model.ConstructSdsSecretConfig(res.GetResourceName(), opts.push.Mesh.SdsUdsPath))
+
+			tlsContext.CommonTlsContext.ValidationContextType = &auth.CommonTlsContext_CombinedValidationContext{
+				CombinedValidationContext: &auth.CommonTlsContext_CombinedCertificateValidationContext{
+					DefaultValidationContext:         &auth.CertificateValidationContext{MatchSubjectAltNames: util.StringToExactMatch(tls.SubjectAltNames)},
+					ValidationContextSdsSecretConfig: authn_model.ConstructSdsSecretConfig(tls.PrivateKey, opts.push.Mesh.SdsUdsPath),
```
Comment (ramaraochavali) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:1043]: Should n't this be  `res.GetRouteResourceName` instead of `tls.PrivateKey` ?

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:1043]: Good point - clear indication I didn't add enough tests

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:1043]: fixed this + added test

**Comment by ramaraochavali** [comment]
@JimmyCYJ Could you please review when you get a chance? There is a follow-up work that needs to happen once this is merged.

**Comment by ramaraochavali** [comment]
@JimmyCYJ @istio/wg-security-maintainers Gentle Ping - Can you PTAL?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 4]
```diff
@@ -251,18 +253,26 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
 	// the files under the well known path.
 	sdsFromFile := false
 	var err error
+
 	if connKey.ResourceName == RootCertReqResourceName && sc.rootCertificateExist(sc.existingRootCertFile) {
 		sdsFromFile = true
 		ns, err = sc.generateRootCertFromExistingFile(sc.existingRootCertFile, token, connKey)
-	} else if connKey.ResourceName == WorkloadKeyCertResourceName &&
-		sc.keyCertificateExist(sc.existingCertChainFile, sc.existingKeyFile) {
+	} else if cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName); ok && cfg.CaCertificatePath != "" {
```
Comment (myidpt) [inline security/pkg/nodeagent/cache/secretcache.go:260]: Nit: I would add some comment for these conditions. I need to read carefully to see the difference between CaCertificatePath and CertificatePath.

Comment (ramaraochavali) [reply security/pkg/nodeagent/cache/secretcache.go:260]: Sure. Will add in follow-up

**Comment by rmichela** [comment]
@howardjohn Can this be backported to 1.6, or if possible, even 1.5.x?

**Pull request merged**